index/segment/mem/segment.go

```diff
@@ -22,224 +22,246 @@ package mem
-	opts      Options
 	offset    int
 	newUUIDFn util.NewUUIDFn
 
-	// Internal state of the segment. The allowed transitions are:
-	//   - Open -> Sealed -> Closed
-	//   - Open -> Closed
 	state struct {
 		sync.RWMutex
-
-		sealed bool
 		closed bool
 	}
 
-	// Mapping of postings list ID to document.
+	// Mapping of postings ID to document.
 	docs struct {
 		sync.RWMutex
-
 		data []doc.Document
 	}
 
-	// Current writer and reader IDs. Writers increment the writer ID for each new
-	// document and only increment the reader ID after the document has been fully
-	// indexed by the segment. Readers do not need to acquire a lock since the writers
-	// only increment the readerID after they have finished inserting a new document.
-	// TODO(jeromefroe): Reevaluate using a big lock here which only permits a single
-	// writer at a time.
-	ids struct {
-		sync.RWMutex
+	// Mapping of term to postings list.
+	termsDict termsDictionary
 
-		writer, reader postings.AtomicID
+	writer struct {
+		sync.Mutex
+		nextID postings.ID
 	}
-
-	// Mapping of field (name and value) to postings list.
-	termsDict termsDictionary
+	readerID postings.AtomicID
 }
 
 // NewSegment returns a new in-memory mutable segment. It will start assigning
-// postings IDs at offset+1.
+// postings IDs at the provided offset.
 func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
 	s := &segment{
-		opts:      opts,
-		offset:    int(offset) + 1, // The first ID assigned by the segment is offset + 1.
+		offset:    int(offset),
 		newUUIDFn: opts.NewUUIDFn(),
 		termsDict: newTermsDict(opts),
+		readerID:  postings.NewAtomicID(offset),
 	}
 
 	s.docs.data = make([]doc.Document, opts.InitialCapacity())
 
-	s.ids.writer = postings.NewAtomicID(offset)
-	s.ids.reader = postings.NewAtomicID(offset)
+	s.writer.nextID = offset
 	return s, nil
 }
 
-func (s *segment) Insert(d doc.Document) error {
+func (s *segment) Insert(d doc.Document) ([]byte, error) {
 	s.state.RLock()
 	defer s.state.RUnlock()
 	if s.state.closed {
-		return sgmt.ErrClosed
+		return nil, sgmt.ErrClosed
 	}
 
-	if s.state.sealed {
-		return errSegmentSealed
+	ds := []doc.Document{d}
+	if err := s.prepareDocs(ds); err != nil {
+		return nil, err
 	}
 
-	docID, err := d.Validate()
-	if err != nil {
-		return err
-	}
+	// Update the document in case we generated a UUID for it.
+	d = ds[0]
 
-	if docID != nil {
-		ok, err := s.termsDict.ContainsTerm(doc.IDReservedFieldName, docID)
-		if err != nil {
-			return fmt.Errorf("could not lookup document ID '%s'", docID)
-		}
-		if ok {
-			// Index already contains this document.
-			return nil
-		}
-	} else {
-		uuid, err := s.newUUIDFn()
-		if err != nil {
-			return err
-		}
-		d.Fields = append(d.Fields, doc.Field{
-			Name:  doc.IDReservedFieldName,
-			Value: uuid,
-		})
-	}
-
-	// TODO: Consider supporting concurrent writes by relaxing the requirement that
-	// inserted documents are immediately searchable.
-	s.ids.Lock()
+	{
```

Comment:
hahah nice

Reply:
Might be overkill, but I think clearly indicating the scope of the lock visually is helpful :)